src/main/java/com/example/Ecommerce/service/VoucherServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+package com.example.Ecommerce.service;
+
+import com.example.Ecommerce.dto.VoucherDto;
+import com.example.Ecommerce.entity.Product;
+import com.example.Ecommerce.entity.Voucher;
+import com.example.Ecommerce.exception.ResourceNotFoundException;
+import com.example.Ecommerce.mapper.VoucherMapper;
+import com.example.Ecommerce.repository.ProductRepo;
+import com.example.Ecommerce.repository.VoucherRepo;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import static com.example.Ecommerce.util.DateUtil.convertToLocalDateTime;
+
+@Service
+@RequiredArgsConstructor
+public class VoucherServiceImpl implements VoucherService {
+
+    private final VoucherRepo voucherRepo;
+    private final ProductRepo productRepo;
+
+
+    @Override
+    public VoucherDto createVoucher(VoucherDto voucherDto) {
+
+        if (voucherRepo.findByCode(voucherDto.getCode()) != null) {
+            throw new ResourceNotFoundException("Voucher with code already exists: " + voucherDto.getCode());
+        }
+        checkIfVoucherIsExpired(voucherDto.getExpiryDate());
```

Comment:
Do the validation - if possible - before any other operation. this concept is called "fail fast"

+ I would name the function: validateVoucher and inside it, it can call all other validation functions which are necessary for creation